### ISSUE TITLE
Mirrors cause wrong cursor position in nested snippets

### ISSUE THREAD

**Issue by JeWe37** [SirVer/ultisnips#1339]
**Expected behavior:**
The cursor should be advanced according to both the characters typed and the characters inside the mirrors of the current tabstop that precede it. 

**Actual behavior:**
The cursor is advanced additionally according to all mirrors, rather than just those preceding the text. This only happens inside nested snippets though and works fine at top-level. It is also noteworthy that this only seems to happen if the mirrors are not separated by any characters between them. If say a "b" is inserted between the two `$1` it works fine.

**Steps to reproduce**

1. Register the following snippets **OR** clone my reproduction fork linked below and run it with Docker
   ```
   snippet test "test"
   $1$1 $0
   endsnippet

   snippet outer "outer"
   $1 $0
   endsnippet
   ```
2. Open a new file with the snippets active and type `outer<tab>test<tab>a`
3. Observe that the cursor has moved behind both "a" characters that appeared, rather than having been placed between them.
-----

- **Operating System**: Manjaro 21.0 Ornara
- **Vim Version**: 
   NVIM v0.4.4
   Build type: Release
- **UltiSnips Version**: 3.2
- **Python inside Vim**: 
   3.9.2 (default, Feb 20 2021, 18:40:11)                                                                                                                                                                                              
   [GCC 10.2.0]
- **Docker repo/vimrc**: [reproduction fork](https://github.com/JeWe37/ultisnips)

**Comment by SirVer** [comment]
Confirmed, it also does not help to be explicit about the tabstop:

```
snippet test "test"
${1}$1 $0
endsnippet

snippet outer "outer"
$1 $0
endsnippet
```

This is probably a off-by-one error or a misguess in the edit actions the user took, probably fairly hard to fix.

**Comment by JeWe37** [comment]
Ok so I think I got the problem tracked down: The cursor is moved properly with mirrors in place by using a dummy NoneditableTextObject that is moved accordingly. The problem is that this is not actually created at the level of the hierarchy at which the user is editing, no, it is created all the way at top level. I fixed it by simply moving the cursor movement into the loop(and caching the current cursor position so vim doesn't get a million updates)! I gotta figure out how to run the test suite and check this doesn't actually cause trouble elsewhere, but if that works find I'm making a pull request later(with some updates to the debug code too BTW, it seems it hasn't been used much since python 2)
